File: examples/ch13/scripts/hello.js
#!/usr/bin/env node
/**
 * hello.js - Node.js 示例脚本
 * 
 * 功能: 接收命令行参数并输出信息
 * 用法: node hello.js [arg1] [arg2] ...
 * 安全提示: 仅在项目目录内操作
 */

// 输出脚本信息
console.log("========================================");
console.log("  Node.js 脚本示例");
console.log("========================================");
console.log();

// 获取 Node.js 版本
console.log(`[Node.js] 版本: ${process.version}`);
console.log(`[Node.js] 平台: ${process.platform}`);
console.log(`[Node.js] 架构: ${process.arch}`);
console.log();

// 处理命令行参数
// process.argv 结构: [node路径, 脚本路径, 用户参数...]
const args = process.argv.slice(2);

console.log(`[Node.js] 参数数量: ${args.length}`);
console.log();

if (args.length > 0) {
    console.log("[Node.js] 命令行参数:");
    args.forEach((arg, index) => {
        console.log(`  argv[${index}] = "${arg}"`);
    });
} else {
    console.log("[Node.js] 没有命令行参数");
}
console.log();

// 字符串操作演示
console.log("[Node.js] 字符串演示:");
const testStr = "Hello, World!";
console.log(`  原始字符串: ${testStr}`);
console.log(`  长度: ${testStr.length}`);
console.log(`  大写: ${testStr.toUpperCase()}`);
console.log(`  小写: ${testStr.toLowerCase()}`);
console.log(`  子串(0,5): ${testStr.substring(0, 5)}`);
console.log(`  包含'World': ${testStr.includes('World')}`);
console.log();

// 数组操作演示
console.log("[Node.js] 数组操作:");
const fruits = ["apple", "banana", "cherry", "date", "elderberry"];
console.log("  水果列表:");
fruits.forEach((fruit, index) => {
    console.log(`    ${index + 1}. ${fruit}`);
});
console.log(`  数量: ${fruits.length}`);
console.log(`  反转: ${[...fruits].reverse().join(', ')}`);
console.log(`  排序: ${[...fruits].sort().join(', ')}`);
console.log();

// 数学计算演示
console.log("[Node.js] 数学计算:");
console.log(`  圆周率: ${Math.PI}`);
console.log(`  2^10 = ${Math.pow(2, 10)}`);
console.log(`  sqrt(144) = ${Math.sqrt(144)}`);
console.log(`  floor(3.7) = ${Math.floor(3.7)}`);
console.log(`  ceil(3.2) = ${Math.ceil(3.2)}`);
console.log();

// 日期时间
const now = new Date();
console.log(`[Node.js] 当前时间: ${now.toISOString()}`);
console.log(`[Node.js] 本地时间: ${now.toLocaleString('zh-CN')}`);
console.log();

console.log("========================================");
console.log("  脚本执行完成");
console.log("========================================");

// 通过 process.exit 设置退出码 (0 = 成功)
process.exit(0);

File: examples/ch13/scripts/hello.lua
#!/usr/bin/env luajit
--[[
hello.lua - LuaJIT 示例脚本

功能: 接收命令行参数并输出信息
用法: luajit hello.lua [arg1] [arg2] ...
安全提示: 仅在项目目录内操作
--]]

-- 输出脚本信息
print("========================================")
print("  LuaJIT 脚本示例")
print("========================================")
print()

-- 获取 Lua 版本
print("[Lua] 版本: " .. (_VERSION or "unknown"))

-- 处理参数
print("[Lua] 参数数量: " .. #arg)
print()

if #arg > 0 then
    print("[Lua] 命令行参数:")
    for i, v in ipairs(arg) do
        print(string.format("  arg[%d] = \"%s\"", i, v))
    end
else
    print("[Lua] 没有命令行参数")
end
print()

-- 字符串操作演示
print("[Lua] 字符串演示:")
local test_str = "Hello, World!"
print("  原始字符串: " .. test_str)
print("  长度: " .. #test_str)
print("  大写: " .. string.upper(test_str))
print("  小写: " .. string.lower(test_str))
print("  子串(1,5): " .. string.sub(test_str, 1, 5))
print()

-- 数学计算演示
print("[Lua] 数学计算:")
print("  圆周率: " .. math.pi)
print("  2^10 = " .. 2^10)
print("  sqrt(144) = " .. math.sqrt(144))
print("  floor(3.7) = " .. math.floor(3.7))
print("  ceil(3.2) = " .. math.ceil(3.2))
print()

-- 表操作演示
print("[Lua] 表操作:")
local fruits = {"apple", "banana", "cherry", "date", "elderberry"}
print("  水果列表:")
for i, fruit in ipairs(fruits) do
    print(string.format("    %d. %s", i, fruit))
end
print("  数量: " .. #fruits)
print()

-- os.date 获取当前时间
print("[Lua] 当前时间: " .. os.date("%Y-%m-%d %H:%M:%S"))
print()

print("========================================")
print("  脚本执行完成")
print("========================================")

-- 通过 os.exit 返回退出码 (0 = 成功)
os.exit(0)
